src/README.rst

```diff
@@ -0,0 +1,28 @@
+``llvm-cov``
+============
+
+We make extensive use of weak pointers, which means that before using them we should always check that they still hold a valid reference to an object.
+In other words, we should do something like:
+
+.. code-block:: c++
+
+   auto sharedPtr = weakPtr.lock();
+
+   if (sharedPtr != nullptr) {
+       ...
+   }
+
+However, in the libCellML codebase, 99.9% of those tests always evaluate to ``true``, which means that ``llvm-cov`` will complain about the ``false`` never being reached while it is just us being "careful" by testing for ``sharedPtr != nullptr``.
+One way to address this problem is to use a C macro:
+
+.. code-block:: c++
+
+   #ifdef LLVM_COVERAGE
+   #    define VALID_SHARED_PTR(ptr) (true)
+   #else
+   #    define VALID_SHARED_PTR(ptr) (ptr != nullptr)
+   #endif
+
+It works fine, but... it's a C macro and we really don't want to go that route, not to mention that it relies on conditional compilation which is not great either.
+Yet, we want 100% branch coverage with ``llvm-cov``.
+So, instead, we have decided not to test for ``sharedPtr != nullptr``.
```

Comment:
Not a show-stopped for this PR, but this seems like information that should be in the developer guides not the repository README...and could be much less verbose.

Reply:
Developer guide... which most people don't read? I don't even know / remember where it is.

I agree that it's a bit verbose and that's the reason it's in a dedicated `README.rst` file where I thought I would put as much information as possible so that the next time we go through the code we don't end up scratching our heads about why we did things the way we did them.

Reply:
I'm with @nickerso  on this, this is not the place for this information.  The developer guides come from the `docs` directory, literally one directory away from where this is now.  I'm sure you can find it.

Reply:
My bad, I had clearly never realised that they were in the same repo. Addressed in 426cefe0.